mlflow/java/src/main/java/org/mlflow/sagemaker/ScoringServer.java

```diff
@@ -0,0 +1,291 @@
+  }
+
+  private static Predictor loadPredictorFromPath(String modelPath)
+      throws PredictorLoadingException {
+    Model config = null;
+    try {
+      config = Model.fromRootPath(modelPath);
+    } catch (IOException e) {
+      throw new PredictorLoadingException(
+          "Failed to load the configuration for the MLFlow model at the specified path.");
+    }
+    return (new MLeapLoader()).load(config);
+  }
+
+  /**
+   * Starts the scoring server locally on a randomly-selected, available port
+   *
+   * @throws IllegalStateException If the server is already active on another port
+   * @throws ServerStateChangeException If the server failed to start and was inactive prior to the
+   *     invocation of this method
+   */
+  public void start() {
+    // Setting port zero instructs Jetty to select a random port
+    start(0);
+  }
+
+  /**
+   * Starts the scoring server locally on the specified port
+   *
+   * @throws IllegalStateException If the server is already active on another port
+   * @throws ServerStateChangeException If the server failed to start and was inactive prior to the
+   *     invocation of this method
+   */
+  public void start(int portNumber) {
+    if (isActive()) {
+      int activePort = this.httpConnector.getLocalPort();
+      throw new IllegalStateException(
+          String.format(
+              "Attempted to start a server that is already active on port %d", activePort));
+    }
+
+    this.httpConnector.setPort(portNumber);
+    try {
+      this.server.start();
+    } catch (Exception e) {
+      throw new ServerStateChangeException(e);
+    }
+    logger.info(String.format("Started scoring server on port: %d", portNumber));
+  }
+
+  /**
+   * Stops the scoring server
+   *
+   * @throws IllegalStateException If the server is already active on another port
+   * @throws ServerStateChangeException If the server failed to start and was inactive prior to the
+   *     invocation of this method
+   */
+  public void stop() {
+    try {
+      this.server.stop();
+      this.server.join();
+    } catch (Exception e) {
+      throw new ServerStateChangeException(e);
+    }
+    logger.info("Stopped the scoring server successfully.");
+  }
+
+  /** @return `true` if the server is active (running), `false` otherwise */
+  public boolean isActive() {
+    return this.server.isStarted();
+  }
+
+  /**
+   * @return Optional that either: - Contains the port on which the server is running, if the server
+   *     is active - Is empty, if the server is not active
+   */
+  public Optional<Integer> getPort() {
+    int boundPort = this.httpConnector.getLocalPort();
+    if (boundPort >= 0) {
+      return Optional.of(boundPort);
+    } else {
+      // The server connector port request returned an error code
+      return Optional.empty();
+    }
+  }
+
+  public static class ServerStateChangeException extends RuntimeException {
+    ServerStateChangeException(Exception e) {
+      super(e.getMessage());
```

Comment:
Why `e.getMessage()`? Shouldn't we pass e wholesale on to the parent, as the cause? Otherwise we lose the class name and stack trace.

Reply:
Good point. Changed to `super(e)`